File: pubmed.py
import os
import sys
from tqdm import tqdm
import pandas as pd
import numpy as np

# Grab all the PubMed IDs for a particular keyword and check headlines/abstracts for gene names
# Must have EDirect installed on command line and api key specified in .zshrc

term = input("Search PubMed for term: ")
term = "+".join(term.split(" "))
term = term.lower()

if os.path.isfile(f"data/{term}_pubmed_results.txt"): # Results already exist for term and will be overwritten.
	overwrite_permission = input(f"\"{term}\" has saved results that appear to be final. Overwrite? (y/n): ")
	if overwrite_permission != "y":
		sys.exit()

print(f"Searching \"{term}\"...")

gene_names = pd.read_csv("data/ensembl.tsv", sep = "\t")
gene_names = gene_names.apply(lambda x: x.str.lower())
gene_names = dict(zip(gene_names.iloc[:,0], gene_names.iloc[:,1]))
genes = pd.read_csv("data/HuRI_genenames.tsv", sep = "\t")
genes = list(set(list(genes.iloc[:,0])+list(genes.iloc[:,1])))
for g in range(len(genes)):
	genes[g] = genes[g].lower()
genes.sort()
hits = np.zeros(len(genes)).astype(int)
num_articles = 0

# Get articles
for year in range(2000,2026):
	data = os.popen(f"esearch -db pubmed -query \"{term}\" -mindate {year} -maxdate {year} | efetch -format abstract").read()
	data = data.split("\n\n\n")
	data = np.array(data)
	data = np.char.replace(data, " \n", " ")
	data = np.char.replace(data, "\n ", " ")
	data = np.char.replace(data, "\n", "")
	data = np.char.lower(data)

	print(f"Analyzing {len(data)} articles from {year}")
	num_articles += len(data)

	for g in tqdm(range(len(genes))):
		hits[g] += np.logical_or.reduce(((np.char.find(data, f"{genes[g]} expression") >= 0), (np.char.find(data, f"{genes[g]} gene") >= 0), (np.char.find(data, f"gene {genes[g]}") >= 0), (np.char.find(data, gene_names[genes[g]]) >= 0))).sum()

print(f"{num_articles} articles analyzed")
results = pd.DataFrame({"A": genes, "B": hits})
results = results[results.B>0]
if len(results) > 0:
	results.A = results.A.apply(lambda x: x.upper())
	results.to_csv(f"data/{term}_results.tsv", index = False, header = False, sep = "\t")
else:
	print("Analysis finished with no matches")

File: pubmed.jl
function analyze(filename, genes, genenames)
	data = read("data/articles/$(filename)", String) |> 
				uppercase |> 
				x -> split(x, "\n\n\n") |>
				x -> map(y -> join(split(y, "\n\n")[2:end-1], " "), x) |>
				x -> map(y -> replace(y, " \n"=>" "), x) |>
				x -> map(y -> replace(y, "\n "=>" "), x) |>
				x -> map(y -> replace(y, "\n"=>" "), x) |>
				x -> filter(y -> length(y) >= 200, x)

	gene_mask = zeros(Bool, length(genes))
	genename_mask = zeros(Bool, length(genes))
	for g in 1:length(genes)
		for d in 1:length(data)
			if occursin(genes[g], data[d])
				gene_mask[g] = true
				break
			end
		end
	end
	for g in 1:length(genes)
		for d in 1:length(data)
			if occursin(genenames[g], data[d])
				genename_mask[g] = true
				break
			end
		end
	end

	hits = zeros(Int, length(genes))

	# Check instances with both
	for g in collect(1:length(genes))[gene_mask .& genename_mask]
		hits[g] += (map(y -> occursin(" GENE "*genes[g]*" ", y), data) .|| 
			map(y -> occursin(" GENE, "*genes[g]*" ", y), data) .|| 
			map(y -> occursin(" "*genes[g]*" GENE ", y), data) .||
			map(y -> occursin(" "*genes[g]*" EXPRESSION ", y), data) .||
			map(y -> occursin(" "*genenames[g]*" ", y), data)) |> sum
	end

	for g in collect(1:length(genes))[gene_mask .& .~genename_mask]
		hits[g] += (map(y -> occursin(" GENE "*genes[g]*" ", y), data) .|| 
			map(y -> occursin(" GENE, "*genes[g]*" ", y), data) .|| 
			map(y -> occursin(" "*genes[g]*" GENE ", y), data) .||
			map(y -> occursin(" "*genes[g]*" EXPRESSION ", y), data)) |> sum
	end

	for g in collect(1:length(genes))[.~gene_mask .& genename_mask]
		hits[g] += map(y -> occursin(" "*genenames[g]*" ", y), data) |> sum
	end

	return hits, length(data)
end

function analyze_set(disease_name)
	if map(x -> isfile("data/articles/pubmed_$(disease_name)_$(x).txt"), collect(2000:2025)) |> all

		genes = read("data/HuRI_genenames.tsv", String) |> 
					x -> split(x, r"[\n\t]") |> 
					unique |> 
					x -> filter(y -> y != "", x) |> 
					sort

		genenames = read("data/ensembl.tsv", String) |>
					x -> split(x, r"[\n\t]") |>
					x -> Dict(zip(x[4:3:end], x[5:3:end])) |>
					x -> [uppercase(x[g]) for g in genes]

		hits_total = zeros(Int, length(genes))
		n_articles_total = 0

		for year in 2000:2025
			println("Analyzing abstracts from $(year)...")
			hits, n_articles = analyze("pubmed_$(disease_name)_$(year).txt", genes, genenames)
			hits_total += hits
			n_articles_total += n_articles
		end

		output = ""
		for i in collect(1:length(genes))[hits .> 0]
			output = output * genes[i] * "\t" * "$(hits[i])" * "\n"
		end
		
		file = open("data/results/pubmed_$(disease_name)_results.txt", "w")
		write(file, output)
		close(file)

		println("Finished analysis of \"$(disease_name)\" with $(n_articles_total) articles")
	else
		println("Missing files with name data/articles/pubmed_$(disease_name)_YYYY.txt")
	end
end

print("\n\n\nEnter search term as it appears in the .txt files: ")
analyze_set(readline())


